# The Greek Anthology's project : our API

The purpose of this notebook is to document the API of the Greek Anthology's project. It is available at http://anthologiagraeca.org/api.
    
## Access the API

We will start by importing different useful libraries and define our first variables for the requests.

In [1]:
import requests
import json

url = 'http://anthologiagraeca.org/api'
parameters = {
    'format':'json'
}

data = requests.get(url, parameters).json()

The variable `data` gives us, in json format, the available endpoints

In [2]:
data

{'passages': 'https://anthologiagraeca.org/api/passages/?format=json',
 'books': 'https://anthologiagraeca.org/api/books/?format=json',
 'scholia': 'https://anthologiagraeca.org/api/scholia/?format=json',
 'texts': 'https://anthologiagraeca.org/api/texts/?format=json',
 'media': 'https://anthologiagraeca.org/api/media/?format=json',
 'manuscripts': 'https://anthologiagraeca.org/api/manuscripts/?format=json',
 'keywords': 'https://anthologiagraeca.org/api/keywords/?format=json',
 'keyword_categories': 'https://anthologiagraeca.org/api/keyword_categories/?format=json',
 'languages': 'https://anthologiagraeca.org/api/languages/?format=json',
 'descriptions': 'https://anthologiagraeca.org/api/descriptions/?format=json',
 'cities': 'https://anthologiagraeca.org/api/cities/?format=json',
 'authors': 'https://anthologiagraeca.org/api/authors/?format=json',
 'comments': 'https://anthologiagraeca.org/api/comments/?format=json',
 'editions': 'https://anthologiagraeca.org/api/editions/?format=jso

## The endpoint *passages*

The first endpoint (`passages`) is the most important : it contains a list of all the epigrams. Let us have a look at it. 

With the variable "epigrams", we directly make a request to the API (its URL is built thanks to the variables defined above and below).

In [4]:
epigrams = '/passages'
epigrams_res = requests.get(url+epigrams,parameters).json()

The result of our request provides us with a lot information related to the passages, as seen below : 

In [5]:
epigrams_res

{'count': 4134,
 'next': 'https://anthologiagraeca.org/api/passages/?format=json&page=2',
 'previous': None,
 'results': [{'id': 424,
   'book': {'url': 'https://anthologiagraeca.org/api/books/9/?format=json',
    'number': 1},
   'fragment': 1,
   'sub_fragment': '',
   'urn': 'urn:cts:greekLit:tlg7000.tlg001.ag:1.1',
   'url': 'https://anthologiagraeca.org/api/passages/urn:cts:greekLit:tlg7000.tlg001.ag:1.1/?format=json',
   'web_url': '/passages/urn:cts:greekLit:tlg7000.tlg001.ag:1.1/',
   'manuscripts': ['http://digi.ub.uni-heidelberg.de/iiif/2/cpgraec23%3A049.jpg/pct:11.132686084142394,13.202832153269469,60.129449838187696,6.205747605164515/full/0/default.jpg'],
   'texts': [{'language': 'grc',
     'text': '\n                      ἃς οἱ πλάνοι καθεῖλον ἐνθάδ᾽ εἰκόνας\n ἄνακτες ἐστήλωσαν εὐσεβεῖς πάλιν.\n'},
    {'language': 'eng',
     'text': 'Inscribed on the Tabernacle of Saint Sophia\n\nThe images that the hereties took down here our pious sovereigns replaced. '},
    {'langu

As you can see, we have for now 4134 epigrams - the value of `count`.

In [8]:
epigrams_res['count']

4134

## Pagination

As you might have noticed, all of the information are not displayed on the block above: the list is pagined and one can navigate the pages using the values of `next` and `previous`.

By default, each pages has: 

In [9]:
len(epigrams_res['results'])

50

50 epigrams. You can change this value with the parameter `limit`.

In [11]:
parameters.update({'limit':100})

epigrams_res = requests.get(url+epigrams,parameters).json()

len(epigrams_res['results'])

100

This list can be filtered : 
- by book (rendering only the epigrams belonging to one particular book) 
- by author's main name (their English name)
- by keyword id (which can be found on the platform : *e.g.* https://anthologiagraeca.org/keywords/1/ is the URL for the keyword "Elegiac couplet" ; the id is "1". 

> For example: https://anthologiagraeca.org/api/passages/?book__number=5&author__main_name=Meleager&keyword__number=1 will give all the epigrams written by Meleager belonging to book 5 and described with the keyword `1` (Elegiac couplet).


In this next cellule, `alldata` contains all the passages' data for all of our 4134 epigrams! 

In [12]:
url = 'http://anthologiagraeca.org/api/passages'
results = []
pagination = True
while pagination == True :
    alldata = requests.get(url, parameters).json()
    for result in alldata['results'] :
        results.append(result)
    if alldata['next'] is None:
        pagination = False
    else:
        url = alldata['next']
        
alldata

{'count': 4134,
 'next': None,
 'previous': 'https://anthologiagraeca.org/api/passages/?format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&format=json&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=100&limit=10

In [13]:
len(results)

4134

## Data about epigrams

Let us now have a look at the data available for each epigram. Most of these data are present in the list of epigrams (`alldata['results']`), but each epigram has its own endpoint, structured on the basis of its book and its number.

The *Greek Anthology* has 16 books, as you can see here:

In [19]:
number_of_books = requests.get('https://anthologiagraeca.org/api/books/').json()['count']

print(number_of_books)

16


An epigram is normally identified by a number (for exemple 1 or 145).

Sometimes there are two or more epigrams for the same number. In these cases letters are used (*e.g.* 132a).

Based on this information the epigram endpoind will be structured as follows:

`/passages/urn:cts:greekLit:tlg7000.tlg001.ag:`+bookNumber`.`+epigramNumber+epigramLetter

> an example from the platform: https://anthologiagraeca.org/passages/urn:cts:greekLit:tlg7000.tlg001.ag:12.132a/

This url is avaiable in the list of epigrams as one can see in the field `url` of each result (let us take the first one here):

In [15]:
epigrams_res['results'][0]['url']

'https://anthologiagraeca.org/api/passages/urn:cts:greekLit:tlg7000.tlg001.ag:1.1/?format=json'

Let us have a look at the epigram 6.13, which means the epigram number 13 of the book 6:

In [16]:
ep6_13 = requests.get('https://anthologiagraeca.org/api/passages/urn:cts:greekLit:tlg7000.tlg001.ag:6.13').json()

ep6_13

{'id': 394,
 'book': {'url': 'https://anthologiagraeca.org/api/books/5/', 'number': 6},
 'fragment': 13,
 'sub_fragment': '',
 'urn': 'urn:cts:greekLit:tlg7000.tlg001.ag:6.13',
 'url': 'https://anthologiagraeca.org/api/passages/urn:cts:greekLit:tlg7000.tlg001.ag:6.13/',
 'web_url': '/passages/urn:cts:greekLit:tlg7000.tlg001.ag:6.13/',
 'manuscripts': ['http://digi.ub.uni-heidelberg.de/iiif/2/cpgraec23%3A142.jpg/pct:10.614886731391586,73.26114119117034,50.355987055016186,4.664723032069971/full/0/default.jpg',
  'http://digi.ub.uni-heidelberg.de/iiif/2/cpgraec23%3A143.jpg/pct:30.37231055564613,13.349681305818653,51.660362990702126,7.613758509494807/full/0/default.jpg'],
 'texts': [{'language': 'grc',
   'text': 'οἱ τρισσοί τοι ταῦτα τὰ δίκτυα θῆκαν ὅμαιμοι,\nἀγρότα Πάν, ἄλλης ἄλλος ἀπ᾽ ἀγρεσίης:\nὧν ἀπὸ μὲν πτηνῶν Πίγρης τάδε, ταῦτα δὲ Δᾶμις\nτετραπόδων, Κλείτωρ δ᾽ ὁ τρίτος εἰναλίων.\nἀνθ᾽ ὧν τῷ μὲν πέμπε δι᾽ ἠέρος εὔστοχον ἄγρην,\nτῷ δὲ διὰ δρυμῶν, τῷ δὲ δι᾽ ἠϊόνων. '},
  {'language': '

In [17]:
ep12_132a = requests.get('https://anthologiagraeca.org/api/passages/urn:cts:greekLit:tlg7000.tlg001.ag:12.132a').json()

ep12_132a

{'id': 3199,
 'book': {'url': 'https://anthologiagraeca.org/api/books/8/', 'number': 12},
 'fragment': 132,
 'sub_fragment': 'a',
 'urn': 'urn:cts:greekLit:tlg7000.tlg001.ag:12.132a',
 'url': 'https://anthologiagraeca.org/api/passages/urn:cts:greekLit:tlg7000.tlg001.ag:12.132a/',
 'web_url': '/passages/urn:cts:greekLit:tlg7000.tlg001.ag:12.132a/',
 'manuscripts': ['http://digi.ub.uni-heidelberg.de/iiif/2/cpgraec23%3A589.jpg/pct:11.78916,41.5646,60.6891,16.87853/full/0/default.jpg'],
 'texts': [{'language': 'grc',
   'text': 'ἆ ψυχὴ βαρύμοχθε, σὺ δ᾽ ἄρτι μὲν ἐκ πυρὸς αἴθῃ,\n ἄρτι δ᾽ ἀναψύχεις, πνεῦμ᾽ ἀναλεξαμένη.\nτί κλαίεις; τὸν ἄτεγκτον ὅτ᾽ ἐν κόλποισιν Ἔρωτα\n ἔτρεφες, οὐκ ᾔδεις ὡς ἐπὶ σοὶ τρέφετο;\n\n                  οὐκ ᾔδεις; νῦν γνῶθι καλῶν ἄλλαγμα τροφείων,\nπῦρ ἅμα καὶ ψυχρὰν δεξαμένη χιόνα.\n αὐτὴ ταῦθ᾽ εἵλου: φέρε τὸν πόνον. ἄξια πάσχεις\nὧν ἔδρας, ὀπτῷ καιομένη μέλιτι.'},
  {'language': 'eng',
   'text': 'O sore-afflicted soul, now thou bumest in the fire and now thou reviv

The epigram's number is in the key `fragment` and the letter (when it has one) in the key `sub_fragment`

### Images of the manuscript (Codex Palatinus 23)

For each epigram, the corresponding iiif images of the manuscript can be found under the key `manuscript` (a high quality digitization of the *codex palatinus 23* is available on [the website of the Palatinate Library of Heidelberg](https://digi.ub.uni-heidelberg.de/diglit/cpgraec23/0079/image,info)). 

For more information about the manuscript and its images, cf. the section "Manuscript Annotation API" in this document. 

In [18]:
ep6_13['manuscripts']

['http://digi.ub.uni-heidelberg.de/iiif/2/cpgraec23%3A142.jpg/pct:10.614886731391586,73.26114119117034,50.355987055016186,4.664723032069971/full/0/default.jpg',
 'http://digi.ub.uni-heidelberg.de/iiif/2/cpgraec23%3A143.jpg/pct:30.37231055564613,13.349681305818653,51.660362990702126,7.613758509494807/full/0/default.jpg']

Two images are associated with epigram 6.13 since it spans two different pages. 

### Texts

Each epigram has a list of texts wich are associated to it. All the epigrams in our database should have at least the greek text. An epigram can have more than one greek editions of the text and a set of translations in different languages :

In [22]:
ep6_13['texts']

[{'language': 'grc',
  'text': 'οἱ τρισσοί τοι ταῦτα τὰ δίκτυα θῆκαν ὅμαιμοι,\nἀγρότα Πάν, ἄλλης ἄλλος ἀπ᾽ ἀγρεσίης:\nὧν ἀπὸ μὲν πτηνῶν Πίγρης τάδε, ταῦτα δὲ Δᾶμις\nτετραπόδων, Κλείτωρ δ᾽ ὁ τρίτος εἰναλίων.\nἀνθ᾽ ὧν τῷ μὲν πέμπε δι᾽ ἠέρος εὔστοχον ἄγρην,\nτῷ δὲ διὰ δρυμῶν, τῷ δὲ δι᾽ ἠϊόνων. '},
 {'language': 'ita',
  'text': 'Tre fratelli ti hanno dedicato queste reti,\nPan cacciatore, ognuna proveniente da una caccia differente.\nPigre queste di uccelli, Damis queste,\ndi bestie feroci, Cleitore, il terzo, di animali marini.\nIn cambio dai una caccia fortunata al primo in aria\nal secondo nei boschi e al terzo sulle rive.'},
 {'language': 'fra',
  'text': "Ces trois frères t'ont dédié ces filets,\nPan chasseur, chacun issu d'une chasse différente.\nPigres celles-ci, d'oiseaux, Damis celle-ci,\nde bêtes féroces, Cléitor, le troisième, d'animaux marins.\nEn échange donne une bonne chasse au premier dans l'air,\nau deuxième dans les bois et au troisième sur les rivages."},
 {'language': 

### Authors


An epigram is almost always associated to one or more authors (since the attributions are often uncertain):

In [23]:
ep6_13['authors']

[{'tlg_id': 'tlg-1458',
  'names': [{'name': 'Leonidas of Tarentum', 'language': 'eng'},
   {'name': 'Λεωνίδας ὁ Ταραντῖνος', 'language': 'grc'},
   {'name': 'Leonida di Taranto', 'language': 'ita'},
   {'name': 'Λεωνίδας ᾿Αλεξανδρεύς', 'language': 'grc'},
   {'name': "Léonidas d'Alexandrie", 'language': 'fra'},
   {'name': 'Leonidas of Alexandria', 'language': 'eng'}]}]

### Keywords

Each epigram can be associated with keywords.

One can have more information about a keyword on its own endpoint, structured as follow : 

`https://anthologiagraeca.org/api/keywords/`+keyword_id 

(the keyword id can be found here at the end of its URL on the platform, *e.g.* https://anthologiagraeca.org/keywords/1/ is the URL for the keyword "Elegiac couplet" ; the id is "1").

In [24]:
ep6_13['keywords']

[{'id': 1,
  'names': [{'name': 'distique élégiaque', 'language': 'fra'},
   {'name': 'distico elegiaco', 'language': 'ita'},
   {'name': 'Elegiac couplet', 'language': 'eng'}],
  'category': [{'name': 'Formes métriques', 'language': 'fra'},
   {'name': 'Metric forms', 'language': 'eng'},
   {'name': 'Metra', 'language': 'lat'},
   {'name': 'Forme metriche', 'language': 'ita'},
   {'name': 'Formas métricas', 'language': 'por'}]},
 {'id': 4,
  'names': [{'name': 'époque hellénistique', 'language': 'fra'},
   {'name': 'epoca ellenistica', 'language': 'ita'},
   {'name': 'hellenistic period', 'language': 'eng'}],
  'category': [{'name': 'Époques', 'language': 'fra'},
   {'name': 'Periods', 'language': 'eng'},
   {'name': 'Tempora', 'language': 'lat'},
   {'name': 'Periodi', 'language': 'ita'},
   {'name': 'Épocas', 'language': 'por'}]},
 {'id': 73,
  'names': [{'name': 'Validé par William', 'language': 'fra'}],
  'category': [{'name': 'Validation', 'language': 'fra'},
   {'name': 'Validat

In [25]:
keyword_1 = requests.get("https://anthologiagraeca.org/api/keywords/1/").json()

In [26]:
keyword_1

{'id': 1,
 'url': 'https://anthologiagraeca.org/api/keywords/1/',
 'category': {'id': 1,
  'url': 'https://anthologiagraeca.org/api/keyword_categories/1/',
  'names': [{'name': 'Formes métriques', 'language': 'fra'},
   {'name': 'Metric forms', 'language': 'eng'},
   {'name': 'Metra', 'language': 'lat'},
   {'name': 'Forme metriche', 'language': 'ita'},
   {'name': 'Formas métricas', 'language': 'por'}]},
 'names': [{'name': 'distique élégiaque', 'language': 'fra'},
  {'name': 'distico elegiaco', 'language': 'ita'},
  {'name': 'Elegiac couplet', 'language': 'eng'}],
 'alternative_urns': [{'urn': 'https://www.wikidata.org/wiki/Q2082412'}]}

keywords are organized in `categories` and each keyword **must** belong to a category.

The list of the categories is here: https://anthologiagraeca.org/api/keyword_categories/

### Cities (places)

An epigram can be associated with some places (cities or countries or whatever can have geocoordinates). Epigram 6.13 has no cities associated. Let us look to another epigram:

In [27]:
ep7_81 = requests.get('https://anthologiagraeca.org/api/passages/urn:cts:greekLit:tlg7000.tlg001.ag:7.81').json()
ep7_81['cities']

[{'id': 45, 'names': [{'name': 'Lindus', 'language': 'eng'}]},
 {'id': 61,
  'names': [{'name': 'Atenas', 'language': 'por'},
   {'name': 'Classical Athens', 'language': 'eng'},
   {'name': 'Antigua Atenas', 'language': 'spa'},
   {'name': 'Athènes', 'language': 'fra'},
   {'name': 'Atene', 'language': 'ita'},
   {'name': 'Athenae antiquae', 'language': 'lat'}]},
 {'id': 73,
  'names': [{'name': 'Sparta', 'language': 'eng'},
   {'name': 'Sparte', 'language': 'fra'},
   {'name': 'Esparta', 'language': 'spa'},
   {'name': 'Sparta', 'language': 'ita'},
   {'name': 'Sparta', 'language': 'lat'},
   {'name': 'Esparta', 'language': 'por'}]},
 {'id': 78,
  'names': [{'name': 'Antigua Corinto', 'language': 'spa'},
   {'name': 'Ancient Corinth', 'language': 'eng'},
   {'name': 'Corinthe antique', 'language': 'fra'},
   {'name': 'Corinto', 'language': 'ita'}]},
 {'id': 89,
  'names': [{'name': 'Mytilene', 'language': 'eng'},
   {'name': 'Mytilène', 'language': 'fra'},
   {'name': 'Mitilene', 'lan

And let us have a look to one of these cities:

In [28]:
Mytilene = requests.get('https://anthologiagraeca.org/api/cities/89').json()
Mytilene

{'url': 'https://anthologiagraeca.org/api/cities/89/',
 'names': [{'name': 'Mytilene', 'language': 'eng'},
  {'name': 'Mytilène', 'language': 'fra'},
  {'name': 'Mitilene', 'language': 'spa'},
  {'name': 'Mitilene', 'language': 'ita'},
  {'name': 'Mytilene', 'language': 'lat'}],
 'alternative_urns': ['https://www.wikidata.org/wiki/Q42295059',
  'https://pleiades.stoa.org/places/550763'],
 'unique_id': 95,
 'longitude': None,
 'latitude': None,
 'descriptions': [],
 'created_at': '2021-09-09T19:09:12.954728Z',
 'updated_at': '2021-09-09T19:09:12.954734Z'}

### Scholia

On the *Codex Palatinus 23*, *scholia* are often associated to epigrams. Those *scholia* are also rendered on the platform and hence on the API. 

In [30]:
ep6_13['scholia']

[{'book': 6,
  'fragment': 13,
  'sub_fragment': '',
  'number': 1,
  'url': '/passages/urn:cts:greekLit:tlg5011.tlg001.sag:6.13.1/'},
 {'book': 6,
  'fragment': 13,
  'sub_fragment': '',
  'number': 2,
  'url': '/passages/urn:cts:greekLit:tlg5011.tlg001.sag:6.13.2/'}]

Each *scholium* is identified by the epigram to which it is a *scholium* plus a number. For exemple the second *scholium* of the epigram 6.13 will be named 6.13.2. 

The urn of a scholium will be structured as follows:

`/passages/urn:cts:greekLit:tlg5011.tlg001.sag:`+bookNumber.+epigramNumber+epigramLetter`.`+scholiumNumber

> For example: `/passages/urn:cts:greekLit:tlg5011.tlg001.sag:6.13.2/`

A scholium contains very similar information as an epigram: texts in different languages, the picture of the manuscript (in iiif), cities, keywords, commentaries.

**Attention** the texts opening a book have been identified as scholia to the epigram 0 of the book. For example 5.0.1 or 5.0.2 or 12.0.1

In [31]:
sc5_0_1 = requests.get('https://anthologiagraeca.org/api/scholia/urn:cts:greekLit:tlg5011.tlg001.sag:5.0.1/').json()
print(sc5_0_1)
sc5_0_2 = requests.get('https://anthologiagraeca.org/api/scholia/urn:cts:greekLit:tlg5011.tlg001.sag:5.0.2/').json()
print(sc5_0_2)
sc12_0_2 = requests.get('https://anthologiagraeca.org/api/scholia/urn:cts:greekLit:tlg5011.tlg001.sag:12.0.2/').json()
print(sc12_0_2)

{'id': 1999, 'url': 'https://anthologiagraeca.org/api/scholia/urn:cts:greekLit:tlg5011.tlg001.sag:5.0.1/', 'urn': 'urn:cts:greekLit:tlg5011.tlg001.sag:5.0.1', 'web_url': '/passages/urn:cts:greekLit:tlg5011.tlg001.sag:5.0.1/', 'manuscripts': ['http://digi.ub.uni-heidelberg.de/iiif/2/cpgraec23%3A087.jpg/pct:9.335021316210499,67.9883147400695,17.828050776662216,8.264462809917363/full/0/default.jpg'], 'texts': [{'language': 'grc', 'text': 'ἀρχή τῶν ἐρωτικὼν ἐπιγραμμάτων'}, {'language': 'por', 'text': 'Início dos epigramas eróticos.'}, {'language': 'fra', 'text': 'Le début des épigrammes érotiques.'}], 'cities': [], 'keywords': [], 'passage': {'book': 5, 'fragment': 0, 'sub_fragment': '', 'url': '/passages/urn:cts:greekLit:tlg7000.tlg001.ag:5.0/'}, 'comments': [{'description': ['Lemmatiste: L.'], 'language': ['fra']}], 'media': []}
{'id': 2000, 'url': 'https://anthologiagraeca.org/api/scholia/urn:cts:greekLit:tlg5011.tlg001.sag:5.0.2/', 'urn': 'urn:cts:greekLit:tlg5011.tlg001.sag:5.0.2', 'w

### Comments

An epigram can have comments, a sort of footnotes about the epigram as a whole of some part of it. These comments can be multilingual and they sometimes contain some markdown markup:

In [32]:
ep6_13['comments']

[{'description': ['# Fresque de Pompéi\n\n Cette épigramme est représentée, selon Gutzwiller, dans une fresque de Pompéi'],
  'language': ['fra']},
 {'description': ["# Imitations de 6.13\n\n C'est vraisemblablement cette pièce de Léonidas qui a inspiré les épigr. 11-16 et 179-187, entre autres 14 (d'Antipater) et 16 (d'Archias), qui n'en sont que des imitations assez serviles. Elle était, en tous cas, une des plus classiques de toute la série; car c'est celle-là qu'on avait gravée sur le mur d'une maison de Pompéi pour servir de légende à une scène où était peinte cette offrande de trois chasseurs; mais il ne subsiste, des trois distiques que comportait l'inscription, que cinq lettres disséminées et les six premières du v. 6; et il a fallu toute la perspicacité de Dilthey pour y reconnaître des fragments de notre épigramme. \n-P. Waltz "],
  'language': ['fra']}]

### External references

An epigram can be linked to some external source. The idea of the project is to develop an edition which can take into account some "weak links" between the anthological material and some other cultural artefacts. This links are subjective and they do not want to be considered as "scientific". They can be link to a pop song, for example.

They are just hyperlinks. For example:

In [33]:
ep7_710 = requests.get('https://anthologiagraeca.org/api/passages/urn:cts:greekLit:tlg7000.tlg001.ag:7.710').json()
ep7_710['external_references']

[{'title': 'Jacques Lazure, Le Blaffart (2018)',
  'url': 'https://opuscules.ca/article-alire?article=202248'},
 {'title': 'Charles Baudelaire, Remords Posthumes (1857)',
  'url': 'https://poesie.webnet.fr/lesgrandsclassiques/Poemes/charles_baudelaire/remords_posthume'},
 {'title': 'Mylène Farmer, Regrets (1991)',
  'url': 'https://www.youtube.com/watch?v=ph6piqBnkgU'}]

### Internal references

Internal references are links between epigrams of the anthology. The relationship is symmetrical. 

Many types of links are possible. An epigram can be the variation of another (and there are different kind of variation) or just have a vague link. 

This is still work in progress ; the type of internal reference will most probably always be "default".

In [34]:
ep6_13['internal_references']

[{'to_passage': 'https://anthologiagraeca.org/api/passages/urn:cts:greekLit:tlg7000.tlg001.ag:6.11/',
  'reference_type': 'Default'},
 {'to_passage': 'https://anthologiagraeca.org/api/passages/urn:cts:greekLit:tlg7000.tlg001.ag:6.14/',
  'reference_type': 'Default'},
 {'to_passage': 'https://anthologiagraeca.org/api/passages/urn:cts:greekLit:tlg7000.tlg001.ag:6.179/',
  'reference_type': 'Default'},
 {'to_passage': 'https://anthologiagraeca.org/api/passages/urn:cts:greekLit:tlg7000.tlg001.ag:6.186/',
  'reference_type': 'Default'},
 {'to_passage': 'https://anthologiagraeca.org/api/passages/urn:cts:greekLit:tlg7000.tlg001.ag:6.187/',
  'reference_type': 'Default'}]

## Wikidata and other identifiers

Almost all the objects in our data are aligned with a wikidata id. It is the case for `authors`, `keywords` and `cities`. The wikidata id can be found in the field `alternative_urns`.

Let us look for an author: 

In [35]:
meleager = requests.get('https://anthologiagraeca.org/api/authors/1').json()

In [36]:
meleager

{'id': 1,
 'url': 'https://anthologiagraeca.org/api/authors/1/',
 'names': [{'name': 'Callimachus', 'language': 'eng'},
  {'name': 'Καλλίμαχος', 'language': 'grc'},
  {'name': 'Callimaque', 'language': 'fra'}],
 'alternative_urns': ['https://www.wikidata.org/wiki/Q192417',
  'http://data.perseus.org/catalog/urn:cts:greekLit:tlg0533'],
 'city_born': {'url': 'https://anthologiagraeca.org/api/cities/1/',
  'names': [{'name': 'Cyrène', 'language': 'fra'},
   {'name': 'Cirene', 'language': 'spa'},
   {'name': 'Cirene', 'language': 'por'},
   {'name': 'Cyrene', 'language': 'eng'},
   {'name': 'Cyrene', 'language': 'lat'},
   {'name': 'Cirene', 'language': 'ita'}],
  'alternative_urns': ['https://anthologia.ecrituresnumeriques.ca/api/v1/cities/3',
   'https://www.wikidata.org/wiki/Q44112',
   'https://pleiades.stoa.org/places/373778'],
  'unique_id': 3,
  'longitude': '32.80799',
  'latitude': '21.86616',
  'descriptions': [],
  'created_at': '2021-04-08T21:27:25.373000Z',
  'updated_at': '20

In [37]:
meleager['alternative_urns']

['https://www.wikidata.org/wiki/Q192417',
 'http://data.perseus.org/catalog/urn:cts:greekLit:tlg0533']

Other identifiers can be found in alternative_urns: for exemple the TLG for authors (look above) and the pleiades id for places:

In [38]:
Mytilene['alternative_urns']

['https://www.wikidata.org/wiki/Q42295059',
 'https://pleiades.stoa.org/places/550763']

## Manuscript Annotation API

We will shortly describe the API developped by the Palatine Library of Heidelberg regarding the annotation of the manuscript. 

All the annotations can be retrieved as follows:

The manuscript has 649 pages. If one wants all the annotation `a` should be 0 and `b` 648. 

**It is only the Codex Palatinus 23 and not the Sup Graec 384!**

Indeed, the Palatine Library of Heidelberg contains the [first 13 books of the Palatine Anthology](https://digi.ub.uni-heidelberg.de/diglit/cpgraec23/0079/image,info) (from pages 49 to 614). Books 14 et 15 (pages 615 to 709) are held at the Bibliothèque Nationale de France under the name [*Parisinus Supplementum Graecum 384*](https://gallica.bnf.fr/ark:/12148/btv1b8470199g). 

Let us look just at some annotations:

In [39]:
def getAnnotations(a,b):
    result = {}
   
    for page in range(a,b):
       
        page = "{:04d}".format(page)
        url = 'https://anno.ub.uni-heidelberg.de/anno/anno?$regex=true&$target=https://digi.ub.uni-heidelberg.de/diglit/cpgraec23/'+page
        r = requests.get(url)
        
        data = r.json()
        result[page] = []
        
        if data['total'] > 0:
            for item in data['first']['items']:
                result[page].append(item)
                

    return result

In [40]:
getAnnotations(590,599)

{'0590': [{'id': 'https://anno.ub.uni-heidelberg.de/anno/anno/I9jkuBuzRb6lJ193QTK80A',
   'type': 'Annotation',
   'title': 'Epigram 11.401 (part 2/2)',
   'collection': 'diglit',
   'body': [{'purpose': 'classifying',
     'source': 'https://d-nb.info/gnd/1231592788',
     'value': 'Pseudo-Lucianus Samosatensis',
     'label': 'Pseudo-Lucianus Samosatensis'},
    {'purpose': 'linking',
     'predicate': 'http://terminology.lido-schema.org/lido00263',
     'source': 'https://anthologiagraeca.org/passages/urn:cts:greekLit:tlg7000.tlg001.ag:11.401/',
     'value': 'Anthologia Graeca Project'},
    {'purpose': 'linking',
     'predicate': 'http://terminology.lido-schema.org/lido00263',
     'source': 'http://data.perseus.org/citations/urn:cts:greekLit:tlg7000.tlg001.perseus-grc4:11.401',
     'value': 'Perseus'}],
   'target': {'scope': 'https://digi.ub.uni-heidelberg.de/diglit/cpgraec23/0590',
    'source': 'https://digi.ub.uni-heidelberg.de/diglit/cpgraec23/0590/_image',
    'selector':